sqld/src/replication/replica/injector/hook.rs

```diff
@@ -0,0 +1,162 @@
+use std::ffi::{c_int, CStr};
+
+use rusqlite::ffi::{libsql_wal as Wal, PgHdr};
+use sqld_libsql_bindings::ffi::types::XWalFrameFn;
+use sqld_libsql_bindings::init_static_wal_method;
+use sqld_libsql_bindings::wal_hook::WalHook;
+
+use crate::replication::frame::FrameBorrowed;
+use crate::replication::WAL_PAGE_SIZE;
+
+use super::headers::Headers;
+use super::FrameBuffer;
+
+// Those are custom error codes returned by the replicator hook.
+pub const LIBSQL_INJECT_FATAL: c_int = 200;
+/// Injection succeeded, left on a open txn state
+pub const LIBSQL_INJECT_OK_TXN: c_int = 201;
+/// Injection succeeded
+pub const LIBSQL_INJECT_OK: c_int = 202;
+
+pub struct InjectorHookCtx {
+    /// shared frame buffer
+    buffer: FrameBuffer,
+    /// currently in a txn
+    is_txn: bool,
+}
+
+impl InjectorHookCtx {
+    pub fn new(buffer: FrameBuffer) -> Self {
+        Self {
+            buffer,
+            is_txn: false,
+        }
+    }
+
+    fn inject_pages(
+        &mut self,
+        sync_flags: i32,
+        orig: XWalFrameFn,
+        wal: *mut Wal,
+    ) -> anyhow::Result<()> {
+        self.is_txn = true;
+        let buffer = self.buffer.lock();
+        let (mut headers, size_after) = make_page_header(buffer.iter().map(|f| &**f));
+
+        let ret = unsafe {
+            orig(
+                wal,
+                WAL_PAGE_SIZE,
+                headers.as_ptr(),
+                size_after,
+                (size_after != 0) as _,
+                sync_flags,
+            )
+        };
+
+        if ret == 0 {
+            debug_assert!(headers.all_applied());
+            if size_after != 0 {
+                self.is_txn = false;
+            }
+            tracing::trace!("applied frame batch");
+
+            Ok(())
+        } else {
+            anyhow::bail!("failed to apply pages");
```

Comment:
do we bother updating `is_txn` here, should we?

Reply:
we don't bother, because this is caught as a fatal replication error anyway, and everything blows up